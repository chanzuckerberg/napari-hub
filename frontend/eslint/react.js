module.exports = {
  extends: ['airbnb/hooks'],

  plugins: ['napari-hub'],

  rules: {
    'napari-hub/no-root-mui-import': 'error',

    /*
      Prop types aren't necessary since we have TypeScript interfaces for
      prop types.
    */
    'react/prop-types': 'off',

    /*
      React automatically adds the import as part of the new JSX transform:
      https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#nextjs
     */
    'react/react-in-jsx-scope': 'off',

    /*
      Prop spreading is very useful as a simple way for passing multiple
      props to a component. However, we should still be careful about passing
      unnecessary props.
    */
    'react/jsx-props-no-spreading': 'off',

    /*
      This rule isn't really needed anymore since we can use ES6 object defaults
      for props. There are also plans to deprecate the use of `defaultProps`:
      https://github.com/reactjs/rfcs/pull/107
    */
    'react/require-default-props': 'off',
  },
};
