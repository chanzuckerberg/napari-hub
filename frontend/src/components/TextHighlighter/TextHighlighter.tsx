import Highlighter from 'react-highlight-words';

export interface Props {
  /**
   * Class to pass to root element.
   */
  className?: string;

  /**
   * Disables text highlighting and renders the text as-is.
   */
  disabled?: boolean;

  /**
   * The text to render.
   */
  children: string;

  /**
   * Words to highlight.
   */
  words?: Array<string | undefined>;
}

/**
 * Renders text with some of the words highlighted.
 */
export function TextHighlighter({
  className,
  children,
  disabled,
  words = [],
}: Props) {
  return (
    <>
      {disabled ? (
        children
      ) : (
        <Highlighter
          className={className}
          highlightClassName="bg-napari-light"
          searchWords={words.filter(Boolean) as string[]}
          textToHighlight={children}
        />
      )}
    </>
  );
}
