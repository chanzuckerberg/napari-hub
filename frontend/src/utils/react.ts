import { ReactElement, ReactNode } from 'react';

/**
 * Checks if a React node is an element. An element is an instantiated
 * component.
 *
 * @param node The React node
 * @returns True if node is an element
 */
export function isReactElement(node: ReactNode): node is ReactElement {
  const element = node as ReactElement | null | undefined;
  return !!(element?.type && element?.props);
}
