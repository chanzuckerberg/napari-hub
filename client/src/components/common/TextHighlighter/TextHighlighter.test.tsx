import { render } from '@testing-library/react';
import { ComponentProps } from 'react';

import { TextHighlighter } from './TextHighlighter';

function renderComponent(
  props: Partial<ComponentProps<typeof TextHighlighter>> = {},
) {
  return render(
    <TextHighlighter words={['world']} {...props}>
      hello world
    </TextHighlighter>,
  );
}

describe('<TextHighlighter />', () => {
  it('should match snapshot', () => {
    const component = renderComponent();
    expect(component.asFragment()).toMatchSnapshot('with highlighting');
  });

  it('should not render highlighting when disabled', () => {
    const component = renderComponent({ disabled: true });
    expect(component.asFragment()).toMatchSnapshot('no highlighting');
  });
});
