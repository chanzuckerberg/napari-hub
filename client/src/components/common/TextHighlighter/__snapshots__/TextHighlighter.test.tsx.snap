// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextHighlighter /> should match snapshot: with highlighting 1`] = `
<DocumentFragment>
  <span>
    <span
      class=""
    >
      hello 
    </span>
    <mark
      class="bg-napari-light "
    >
      world
    </mark>
  </span>
</DocumentFragment>
`;

exports[`<TextHighlighter /> should not render highlighting when disabled: no highlighting 1`] = `
<DocumentFragment>
  hello world
</DocumentFragment>
`;
